Add user projects retrieval

Merseleo opened the pull request:
Closes #84, Closes #85, Closes #82

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a service, which does x -->

I added the implementation for the following calls
- `GetAllProjectsForUser`
- `GetAllArchivedProjectsForUser`
- `GetAllDeletedProjectsForUser`

I implemented these calls together as `GetAllArchivedProjectsForUser` and `GetAllDeletedProjectsForUser` are only special / filtered versions of the general `GetAllProjectsForUser`. That's why their implementation is quite similar and can be combined. 

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

Commit:
feat: add repository implementation to retrieve projects of a certain user

This especially can be used for the following call:
'GetAllProjectsForUser','GetAllArchivedProjectsForUser','GetAllDeletedProjectsForUser'

Commit:
feat: add service implementation to retrieve projects of a certain user

Furthermore, add a helper to check whether the requesting user is authorized to access another's user resources and refactor project services to use helper functions.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepo.kt:
```diff
@@ -37,6 +39,23 @@ interface IProjectTableRepo {
+     *        If set to another value (e.g., [ProjectStatus.PROJECT_STATUS_DELETED]), only projects
+     *        matching that (e.g., deleted) status will be returned.
+     *
+     * @return A list of [Project] objects matching the specified filter where the given user is member of.
```
With the indentation, the references are not correctly resolved, i.e., not seen as references at all.

```suggestion
     * @param statusFilter (optional) Filter to specify which project status the fetched projects should have.
     * By default, [ProjectStatus.PROJECT_STATUS_ACTIVE] is used, which includes both active and active-locked projects.
     * If set to another value (e.g., [ProjectStatus.PROJECT_STATUS_DELETED]), only projects matching that (e.g.,
     * deleted) status will be returned.
     * @return A list of [Project] objects matching the specified filter where the given user is member of.
```

Merseleo replied:
Thank you for the hint, I will update it.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepo.kt:
```diff
@@ -79,4 +98,27 @@ class ProjectTableRepo(
     }
+
+    override suspend fun getUserProjects(userId: UUID, statusFilter: ProjectStatus): List<Project> = db.dbQuery {
+        // TODO (question for reviewer too): Check whether library automatically optimize this and first filter ProjectTable and afterwards joins
```
No, exposed only translates this to an SQL command (you can see it in your logs):
```sql
SELECT project.id, project."name", project.status, project.current_stage, project.max_stage,
project.similarity_threshold, project.snowballing_type, project.review_maybe_allowed,
project.review_decision_matrix, project.fetcher_apis, project.current_stage_started_at, project.created_at,
project.created_by, project.modified_at, project.modified_by, project.deleted_at, project.deleted_by,
project.archived_at, project.archived_by FROM project
INNER JOIN project_member ON project.id = project_member.project_id
WHERE (project_member.user_id = '07b1b48d-0b81-4381-81b4-acdb6ded5197') AND
((project.status = 1) OR (project.status = 2))
```

Merseleo replied:
Ok, not as I hoped. Because when there exist many projects this could be reduce the runtime for the join dramatically if we first filter and afterwards join instead of first join and afterwards filter.

Merseleo replied:
But I didnt found any solution, so maybe Postgres does optimize the parsed SQL code.

Slartibartfass2 replied:
I don't think we need to think about optimizing this right now.

Merseleo replied:
Ok

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepo.kt:
```diff
@@ -79,4 +98,27 @@ class ProjectTableRepo(
+                when (statusFilter) {
+                    ProjectStatus.PROJECT_STATUS_ACTIVE, ProjectStatus.PROJECT_STATUS_ACTIVE_LOCKED ->
+                        (ProjectTable.status eq ProjectStatus.PROJECT_STATUS_ACTIVE) or
+                            (ProjectTable.status eq ProjectStatus.PROJECT_STATUS_ACTIVE_LOCKED)
```
It seems unintuitive to me that when the status is active or active locked that both are filtered, compared to archived or deleted.

What is your opinion on this?:
```kotlin
    override suspend fun getUserProjects(userId: UUID, statusFilters: Set<ProjectStatus>): List<Project> = db.dbQuery {
        val filter = statusFilters.map { ProjectTable.status eq it }.reduceOrNull { acc, expression -> acc or expression }

        if (filter == null) {
            throw Exception("Here we should throw some custom exception")
        }

        (ProjectTable innerJoin ProjectMemberTable)
            .select(ProjectTable.columns)
            .where { ProjectMemberTable.userId eq userId }
            .andWhere { filter }
            .map { it.toProject() }
    }
```

This works for all available statuses, we could further filter out `UNRECOGNIZED` or `UNSPECIFIED`.

We would then call it like this:
```kotlin
repo.getUserProjects(requestedUserId, setOf(ProjectStatus.PROJECT_STATUS_ARCHIVED))
```

Merseleo replied:
In my opinion that doesn't make it more intuitive but even more complicated, but I notice another advantage of your solution: It is easier to evolve this repository function, especially in case we add new statuses or want to (for whatever reason) request all archived and active projects at once. So I like your solution and will update the parameter from a single filter to a set of filter.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -1,26 +1,45 @@
+    suspend fun getAllArchivedProjectsForUser(request: Base.Id): GrpcProject.List
+
+    /**
+     * Service implementation of [SnowballRService.getAllDeletedProjectsForUser]
```
None of these service implementations are connected to the server calls.
```suggestion
     * Service implementation of [SnowballRService.getAllProjectsForUser].
     */
    suspend fun getAllProjectsForUser(request: Base.Id): GrpcProject.List

    /**
     * Service implementation of [SnowballRService.getAllArchivedProjectsForUser].
     */
    suspend fun getAllArchivedProjectsForUser(request: Base.Id): GrpcProject.List

    /**
     * Service implementation of [SnowballRService.getAllDeletedProjectsForUser].
```

Merseleo replied:
I totally forgot this, thank you.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -37,22 +56,48 @@ class ProjectService(
     private val userRepo: IUserTableRepo,
 ) : IProjectService {
-    override suspend fun createProject(request: ProjectOuterClass.Project.Create): ProjectOuterClass.Project {
+    private fun toGrpcProjects(projects: List<Project>): GrpcProject.List {
```
We already have the `.toGrpcProject()` mapper in the `Project` dto. Should we also move this there?

Merseleo replied:
I can move it.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ServiceChecks.kt:
```diff
@@ -18,3 +23,31 @@ fun verifyServerAdminRole(user: User, getException: (String) -> SnowballRExcepti
+ * @param userRepo The user repository used to retrieve user data.
+ * @param accessType The type of access being attempted, used to construct the exception if unauthorized.
+ */
+suspend fun authorizeAccessTo(requestedUserId: UUID, userRepo: IUserTableRepo, accessType: AccessType) {
```
This is a great addition. We use similar checks in the `UserService` do you think we can integrate this method there?

Merseleo replied:
I know that we use it there as well and @domienderle used something similar in his PR, but I first wanted to speak with him about this. We already discussed this and I will refactor this a little bit so we have less duplicate code.

Merseleo replied:
However, this is not a small update or refactoring, so I will do this work in another PR as soon as this one is merged, if you are ok with that.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepoTest.kt:
```diff
@@ -3,23 +3,28 @@ package se.uulm.snowballr.backend.repository
 @ExperimentalCoroutinesApi
 @DelicateCoroutinesApi
-class ProjectTableRepoTest : H2DatabaseTest(arrayOf(ProjectTable), true) {
+class ProjectTableRepoTest : H2DatabaseTest(arrayOf(ProjectTable, ProjectMemberTable, UserTable), true) {
```
Since we already request a test user, we don't need to list the `UserTable`.
```suggestion
class ProjectTableRepoTest : H2DatabaseTest(arrayOf(ProjectTable, ProjectMemberTable), true) {
```

Merseleo replied:
Ok.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepoTest.kt:
```diff
@@ -38,6 +43,26 @@ class ProjectTableRepoTest : H2DatabaseTest(arrayOf(ProjectTable), true) {
     }
 
+    private suspend fun createExampleUser(email: String) = db.dbQuery {
+        UserTable
```
A similar call is used in `ProjectMemberTableRepoTest` and we probably need this more often, let's extract it so that every test class can use it.

Merseleo replied:
Thank you for the hint, I did not know about the `ProjectMemberTableRepoTest`. I will extract it.

Merseleo replied:
However, I just saw that you used the production code to assign a user to a project and did not use the database directly. Is there a reason why? Because I remove this use of the repository function as this is not correct in my opinion. The tests should not rely on the correctness of other repository functions.

Slartibartfass2 replied:
I see your point, and I also think that with this we test if the other repository functions work, i.e., test another function as well.

Merseleo replied:
Ok, so it is ok that I removed this repository call from the test helper and used the direct database connection instead?

Slartibartfass2 replied:
yes

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetAllArchivedProjectsForUserTest.kt:
```diff
@@ -0,0 +1,101 @@
+            }
+        }
+
+    fun `When a user retrieve its own archived projects, then all projects are returned successfully`() =
```
```suggestion
    fun `When a user retrieves its own archived projects, then all projects are returned successfully`() =
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetAllDeletedProjectsForUserTest.kt:
```diff
@@ -0,0 +1,96 @@
+            assertThrows<UnauthorizedException.Single> { mainService.getAllDeletedProjectsForUser(getExampleRequest()) }
+        }
+
+    fun `When a user retrieve its own deleted projects, then all projects are returned successfully`() = testCoroutine {
```
```suggestion
    fun `When a user retrieves its own deleted projects, then all projects are returned successfully`() = testCoroutine {
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetAllProjectsForUserTest.kt:
```diff
@@ -0,0 +1,87 @@
+            assertThrows<UnauthorizedException.Single> { mainService.getAllProjectsForUser(getExampleRequest()) }
+        }
+
+    fun `When a user retrieve its own projects, then all projects are returned successfully`() = testCoroutine {
```
```suggestion
    fun `When a user retrieves its own projects, then all projects are returned successfully`() = testCoroutine {
```

Slartibartfass2 reviewed (changes requested):
Great work! Another 3 new calls, we can add to our collection.

Merseleo committed:
docs: fix typos in test descriptions

Co-authored-by: Felix Schlegel <40503329+Slartibartfass2@users.noreply.github.com>

Commit:
refactor: change the statusFilter parameter of getUserProjects repository function to accept a set of filters

Commit:
feat: delegate the getAll*ProjectsForUsers call from the gRPC layer to the project service

Commit:
refactor: move the toGrpcProjects function into the Project DTO

Commit:
refactor: extract database helper function to create a user and assign this user to a project into a separate RepositoryHelper.kt file in the repository subdirectory.

Additionally, add the decorator @ExperimentalCoroutinesApi and @DelicateCoroutinesApi.

Merseleo reviewed (commented):
I hope I did not forgot anything. I would keep the "inefficient" joining for now, as I did not found any better solution, maybe we will find it later and then we can update this code.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/model/dto/Project.kt:
```diff
@@ -57,3 +57,12 @@ fun Project.toGrpcProject(): ProjectOuterClass.Project {
+    val builder = ProjectOuterClass.Project.List.newBuilder()
+    projects.forEach { builder.addProjects(it.toGrpcProject()) }
+    return builder.build()
+}
```
By using an extension function similar to the above, the call reads more naturally.
```suggestion
/**
 * Creates a list of [ProjectOuterClass.Project]s from this list of [Project]s.
 */
fun List<Project>.toGrpcProjects(): ProjectOuterClass.Project.List {
    val builder = ProjectOuterClass.Project.List.newBuilder()
    this.forEach { builder.addProjects(it.toGrpcProject()) }
    return builder.build()
}
```

This can be call like this:
```kotlin
projects.toGrpcProjects()
// instead of
toGrpcProjects(projects)
```

Merseleo replied:
I have already tried this myself but failed, now I know how to do it, thank you!

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -1,26 +1,45 @@
-import snowballr.ProjectOuterClass
+import snowballr.Base
+import snowballr.ProjectOuterClass.ProjectStatus
+import snowballr.ProjectOuterClass.Project as GrpcProject
```
I like the aliasing.

Merseleo replied:
I agree, it makes the code more readable and we can continue to do this for future code where it is appropriate.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepo.kt:
```diff
@@ -37,6 +40,29 @@ interface IProjectTableRepo {
+     */
+    suspend fun getUserProjects(
+        userId: UUID,
+        statusFilters: Set<ProjectStatus> = setOf(
+            ProjectStatus.PROJECT_STATUS_ACTIVE,
+            ProjectStatus.PROJECT_STATUS_ACTIVE_LOCKED,
+        ),
```
Do we need a default value, or is this just for convenience? Do we anticipate that this call is used with the same values several times?

Merseleo replied:
I don't think it will be used more than once and yes you could leave them out. My idea was to improve the readability and simplify the usage. If you don't specify anything further you always assume an active project and if you want ‘special’ projects so you need to specify that. In my opinion, this makes it easier to use, but I can of course delete them if you want.

Slartibartfass2 replied:
I give you the benefit of doubt.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/RepositoryHelper.kt:
```diff
@@ -0,0 +1,64 @@
+/**
+ * This class acts as a collection of create methods to create database entries for testing purposes.
+ */
+object RepositoryHelper {
```
Are there advantages to having this class in this directory instead of the `repository` sub-directory?

Merseleo replied:
No there aren't really any advantages. I thought we maybe want to use `repository` directory only for the tests and all helpers etc. are in the test directory. 
But I will move it.

Slartibartfass2 reviewed (changes requested):
Great work! Nearly done.

Slartibartfass2 commented:
We need to add the following to the repository helper.
```
@ExperimentalCoroutinesApi
@DelicateCoroutinesApi
```

Merseleo reviewed (commented):
I was so free and did not make the two changes as extra commits but quickly integrated them into the existing commits and also rebased, I hope that was okay.

Slartibartfass2 reviewed (approved):
We're done here. Great work!
